src/backend/distributed/operations/shard_split.c

```diff
@@ -0,0 +1,960 @@
+		 */
+		CreateAndCopySplitShardsForShardGroup(
+			sourceShardToCopyNode,
+			sourceColocatedShardIntervalList,
+			shardGroupSplitIntervalListList,
+			workersForPlacementList);
+
+		/*
+		 * Drop old shards and delete related metadata. Have to do that before
+		 * creating the new shard metadata, because there's cross-checks
+		 * preventing inconsistent metadata (like overlapping shards).
+		 */
+		DropShardList(sourceColocatedShardIntervalList);
+
+		/* Insert new shard and placement metdata */
+		InsertSplitChildrenShardMetadata(shardGroupSplitIntervalListList,
+										 workersForPlacementList);
+
+		/*
+		 * Create foreign keys if exists after the metadata changes happening in
+		 * DropShardList() and InsertSplitChildrenShardMetadata() because the foreign
+		 * key creation depends on the new metadata.
+		 */
+		CreateForeignKeyConstraints(shardGroupSplitIntervalListList,
+									workersForPlacementList);
+	}
+	PG_CATCH();
+	{
+		/* Do a best effort cleanup of shards created on workers in the above block */
+		TryDropSplitShardsOnFailure(shardGroupSplitIntervalListList,
+									workersForPlacementList);
+
+		PG_RE_THROW();
+	}
+	PG_END_TRY();
+
+	CitusInvalidateRelcacheByRelid(DistShardRelationId());
+}
+
+
+/* Create ShardGroup split children on a list of corresponding workers. */
+static void
+CreateSplitShardsForShardGroup(List *shardGroupSplitIntervalListList,
+							   List *workersForPlacementList)
+{
+	/* Iterate on shard interval list for shard group */
+	List *shardIntervalList = NULL;
+	foreach_ptr(shardIntervalList, shardGroupSplitIntervalListList)
+	{
+		/* Iterate on split shard interval list and corresponding placement worker */
+		ShardInterval *shardInterval = NULL;
+		WorkerNode *workerPlacementNode = NULL;
+		forboth_ptr(shardInterval, shardIntervalList, workerPlacementNode,
+					workersForPlacementList)
+		{
+			/* Populate list of commands necessary to create shard interval on destination */
+			List *splitShardCreationCommandList = GetPreLoadTableCreationCommands(
+				shardInterval->relationId,
+				false, /* includeSequenceDefaults */
+				NULL /* auto add columnar options for cstore tables */);
+			splitShardCreationCommandList = WorkerApplyShardDDLCommandList(
+				splitShardCreationCommandList,
+				shardInterval->shardId);
+
+			/* Create new split child shard on the specified placement list */
+			CreateObjectOnPlacement(splitShardCreationCommandList, workerPlacementNode);
+		}
+	}
+}
+
+
+/* Create ShardGroup auxiliary structures (indexes, stats, replicaindentities, triggers)
+ * on a list of corresponding workers.
+ */
+static void
+CreateAuxiliaryStructuresForShardGroup(List *shardGroupSplitIntervalListList,
+									   List *workersForPlacementList)
+{
+	/*
+	 * Create auxiliary structures post copy.
+	 */
+	List *shardIntervalList = NULL;
+	foreach_ptr(shardIntervalList, shardGroupSplitIntervalListList)
```

Comment:
I still find it a bit difficult to remember how shardGroupSplitIntervalListList is structured every time I see it.

Reply:
@marcocitus 
I can 1) Add comment like we did for others 2)  Define an intermediate like below, 

```
typedef struct ShardSplitIntervals
{
 List *shardSplitIntervalList;
} ShardSplitIntervals;

and then return List of ShardSplitIntervals to avoid Lists of List.
```

From our discussion in team's channel :
 `we have a struct, we could also put the source ShardInterval in the struct, and then we don't need to separately pass sourceColocatedShardIntervalList to CreateSplitShardsForShardGroup`

I don't want to add source ShardInterval to the struct as only 1 method where 'sourceColocatedShardIntervalList' actually has a use for it.

Reply:
I think another option here could be to move the inner loop into a separate function, such that the function name communicates what we are doing (CreateAuxiliaryStructuresForSingleShardSplits?)

alternatively, having more descriptive comments can help
```c
/*
 * Iterate over all the shards in the shard group
 */
foreach(..)
...
    /*
     * Create the objects for each split coming out of a single shard in a shard group
     */
```

Reply:
Given I am following this pattern in a couple more places, I have added new comments across to explain better.